### PR TITLE
Use split_parser branch for markdown grammar

### PR DESCRIPTION
**Pull request by MDeiml**
See #3072.

Only a draft because `markdown_inline` grammar is not injected properly, but to me it seems like this is a problem with helix and not the language config? After a few experiments I think the grammar _is_ injected, but highlighting is still not done according to the injected language. I'm not very sure though.

Another weird behavior is with fenced code blocks. They seem to work to some degree, but if you enter something like
~~~markdown
```rust
let foo = "bar";
```
~~~
then `let` is highlighted correctly while `"bar"` is not highlighted at all.

### CODE REVIEW

**Comment by MDeiml** [comment]
Changing this to ready since I found a solution to the problems in #3129.

**Comment by the-mikedavis** [comment]
I tried this branch locally but I don't see any qualitative improvements to performance (we'll need some measurements like from #940 to be definitive). Specifically I see the some slowness with incremental updates while editing (for example typing quickly in insert mode) on a reasonably large and complicated markdown document like [the changelog](https://github.com/helix-editor/helix/blob/5551cfd0ba36b5cda7a49c78631464240213e938/CHANGELOG.md). Is this branch of the grammar(s) more correct than master? This doesn't feel like a regression so I'm certainly not against merging it but I'm curious what the improvement is of splitting the parsers?

**Comment by MDeiml** [comment]
I'm no longer maintaining the master branch, I just kept it because some projects seem to depend on it.

It was not really possible to implement a correct enough markdown parser in a single grammar. The specs also recommend to parse markdown in 2 passes. Now after splitting it into 2 grammars they work closer to that recommendation.

They're more correct now and with some optimizations to incremental parsing I think they'll also be faster. (But that'll be topic of future PRs).

**Comment by MDeiml** [comment]
Specifically I think that tree-sitter reparses all injected ranges if they contain "conflicts" no matter if they changed or not. That means that parsing is linear with the number of injected ranges. I think we could just not reparse ranges if they did not change at all (parsing should be deterministic with regards to the input anyways).

**Comment by the-mikedavis** [comment]
I'm not very familiar with the changes but I think archseer might've covered that when implementing combined injections: https://github.com/helix-editor/helix/compare/6728e4449038e9481b72251441182d508c165a9c...7c9ebd05b83e90e55d032f65d9405ad265b82258

We don't use the stock tree-sitter-cli highlighting code since it isn't incremental

**Comment by MDeiml** [comment]
Just looking at the code that doesn't seem to be the case, but I should do some more investigation.

**Comment by MDeiml** [comment]
It's related to this TODO:
https://github.com/helix-editor/helix/blob/681c0a91dc36f9db1d23bd31ac1b66414d2a6099/helix-core/src/syntax.rs#L724
Just going one step further and also avoid parsing layers that are after the edits

**Comment by the-mikedavis** [comment]
This works well for me locally but let's await further review/merging in #3129

**Comment by archseer** [comment]
Can you resolve the conflict? Looks good to merge otherwise 👍🏻

File: languages.toml [hunk 1]
```diff
@@ -868,7 +868,18 @@ indent = { tab-width = 2, unit = "  " }
 
 [[grammar]]
 name = "markdown"
-source = { git = "https://github.com/MDeiml/tree-sitter-markdown", rev = "ab15701d8f3f68aeb74e30573b7d669a6ef2a7ed" }
+source = { git = "https://github.com/MDeiml/tree-sitter-markdown", rev = "142a5b4a1b092b64c9f5db8f11558f9dd4009a1b", subpath = "tree-sitter-markdown" }
+
+[[language]]
+name = "markdown_inline"
+scope = "source.markdown_inline"
+injection-regex = "markdown_inline"
```
Comment (archseer) [inline languages.toml:876]: How about `markdown.inline`?

Comment (MDeiml) [reply languages.toml:876]: Sure. For all of those? I have to admit I don't really understand what `scope` is for.

Comment (the-mikedavis) [reply languages.toml:876]: > For all of those?

Yeah but `injection-regex` should probably be `"markdown\\.inline"` to be exact.

They're like textmate scopes. Currently we use them for some language naming things like what to send to a Language Server as `language_id`.

**Review by the-mikedavis — APPROVED** [review]
Works great! Thanks for working on this plus going the extra mile to add the new predicate :)

**Pull request merged**